### PR TITLE
Add customer code

### PR DESCRIPTION
**Pull request by vzhang23**

### CODE REVIEW

**Comment by vzhang23** [comment]
@mclhtay can you review my code change, and change the admin page so it is working with my change in database?

File: HotelBooking/Controllers/bill_controller.py [hunk 1]
```diff
@@ -12,5 +13,20 @@ def __init__(self) -> None:
...
+    def get_bill(self, bill_id: int) -> Bill:
+        return self.bill.get_bill(bill_id)
+
     def pay_bill(self, bill_id: int):
         self.bill.pay_bill(bill_id)
+
+    def cancel_bill(self, bill_id: int):
+        self.bill.cancel_bill(bill_id)
+
+    def get_available_bill(self, customer_id) -> List[Bill]:
```
Comment (mclhtay) [inline HotelBooking/Controllers/bill_controller.py]: Can you use plural here to reflect the return type?
`get_available_bills`

File: HotelBooking/Controllers/bill_controller.py [hunk 2]
```diff
@@ -12,5 +13,20 @@ def __init__(self) -> None:
...
+        return self.bill.get_available_bill(customer_id)
+
+    def get_all_bill(self, customer_id) -> List[Bill]:
```
Comment (mclhtay) [inline HotelBooking/Controllers/bill_controller.py]: Plural.

File: HotelBooking/Controllers/reservation_controller.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+from HotelBooking.Models import reservation
+from HotelBooking.Models.bill import Bill
+from HotelBooking.Models.reservation import Reservation
+from HotelBooking.Models.room import Room
+from dateutil import parser
+
+from HotelBooking.Models.roomType import RoomType
+
+# Cost per hour of stay
+ROOM_PRICING = {
```
Comment (mclhtay) [inline HotelBooking/Controllers/reservation_controller.py]: Did you not add the room pricings to the database? If so, use the pricing in the database.

File: HotelBooking/Controllers/reservation_controller.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    room: Room
+    roomType: RoomType
+    reservation: Reservation
+    bill_controller: BillController
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.room = Room()
+        self.bill_controller = BillController()
+        self.roomType = RoomType()
```
Comment (mclhtay) [inline HotelBooking/Controllers/reservation_controller.py]: follow Python name convention of snake_case

File: HotelBooking/Controllers/reservation_controller.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        self.reservation = Reservation()
+
+    def get_open_reservation(self, userID=None) -> List[Reservation]:
```
Comment (mclhtay) [inline HotelBooking/Controllers/reservation_controller.py]: Plural.

Comment (mclhtay) [reply HotelBooking/Controllers/reservation_controller.py]: also instead of userID, consider `user_id` or `customer_id`

File: HotelBooking/Controllers/reservation_controller.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        return self.reservation.get_reservation_with_status("OPEN", userID)
+
+    def get_reservation(self, userID=None) -> List[Reservation]:
```
Comment (mclhtay) [inline HotelBooking/Controllers/reservation_controller.py]: plural, and userId

File: HotelBooking/Controllers/reservation_controller.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        return self.reservation.get_reservation_by_customer_id(userID)
+
+    def get_stay(self, userID=None) -> List[Reservation]:
```
Comment (mclhtay) [inline HotelBooking/Controllers/reservation_controller.py]: this method's name is misleading. What is get_stay?

File: HotelBooking/Controllers/reservation_controller.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        return self.reservation.get_reservation_with_status("CLOSED", userID)
+
+    def reserve_room(self, room_id: str, customer_id: int, startDate, duration: int):
+        self.room.update_room_status(room_id, "RESERVED", customer_id)
+        r = self.room.get_room_by_id(room_id)
+        price = float(self.roomType.get_Price(r.room_type))
```
Comment (mclhtay) [inline HotelBooking/Controllers/reservation_controller.py]: You're using the pricing from the database here to calculate price, so please remove the above pricing map.

Comment (mclhtay) [reply HotelBooking/Controllers/reservation_controller.py]: `start_date`

File: HotelBooking/Controllers/reservation_controller.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        self.reservation.create_reservation(
+            "OPEN", customer_id, room_id, bill.bill_id, startDate, duration)
+
+    def cancel_reservation(self, room_id: str, reservation_id: int):
+        self.room.update_room_status(room_id, "AVAILABLE")
+        self.reservation.update_reservation_status(reservation_id, "CANCELED")
+        self.bill_controller.cancel_bill(
+            self.reservation.get_reservation_by_id(reservation_id).bill_id)
+
+    def modify_reservation_date(self, reservation_id: int, newStartDate: str):
```
Comment (mclhtay) [inline HotelBooking/Controllers/reservation_controller.py]: `new_start_date`

File: HotelBooking/Controllers/reservation_controller.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        self.reservation.update_reservation_start_date(
+            reservation_id, newStartDate)
+        self.bill_controller
+
+    def modify_reservation_duration(self, reservation_id: int, newDuration: int):
```
Comment (mclhtay) [inline HotelBooking/Controllers/reservation_controller.py]: `new_duration`

File: HotelBooking/Controllers/reservation_controller.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        self.reservation.update_reservation_duration(
+            reservation_id, newDuration)
+        currentReservation = self.reservation.get_reservation_by_id(
+            reservation_id)
+        price: float = self.roomType.get_Price(
+            self.room.get_room_by_id(currentReservation.room_id).room_type)
+        self.bill_controller
+        newBillID = self.bill_controller.modify(self.reservation.get_reservation_by_id(
```
Comment (mclhtay) [inline HotelBooking/Controllers/reservation_controller.py]: `new_bill_id`

File: HotelBooking/Controllers/reservation_controller.py [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+    def modify_reservation_date(self, reservation_id: int, newStartDate: str):
+        self.reservation.update_reservation_start_date(
+            reservation_id, newStartDate)
+        self.bill_controller
+
+    def modify_reservation_duration(self, reservation_id: int, newDuration: int):
+        self.reservation.update_reservation_duration(
+            reservation_id, newDuration)
+        currentReservation = self.reservation.get_reservation_by_id(
```
Comment (mclhtay) [inline HotelBooking/Controllers/reservation_controller.py]: `current_reservation`

File: HotelBooking/Controllers/room_controller.py [hunk 13]
```diff
@@ -1,10 +1,16 @@
 from datetime import datetime
+from tracemalloc import start
```
Comment (mclhtay) [inline HotelBooking/Controllers/room_controller.py]: what is this import?

File: HotelBooking/Controllers/room_controller.py [hunk 14]
```diff
@@ -1,10 +1,16 @@
...
+from HotelBooking.Models import reservation
+from HotelBooking.Models.bill import Bill
+from HotelBooking.Models.reservation import Reservation
 from HotelBooking.Models.room import Room
 from dateutil import parser
 
+from HotelBooking.Models.roomType import RoomType
+
 # Cost per hour of stay
 ROOM_PRICING = {
```
Comment (mclhtay) [inline HotelBooking/Controllers/room_controller.py]: remove `ROOM_PRICING` and use the pricing from RoomType to stay consistent.

File: HotelBooking/Controllers/room_controller.py [hunk 15]
```diff
@@ -17,42 +23,38 @@
...
+        self.bill = Bill()
+        self.roomType = RoomType()
+        self.reservation = Reservation()
 
     def get_checked_in_rooms(self) -> List[Room]:
         return self.room.get_checked_in_rooms()
 
-    def get_reserved_rooms(self) -> List[Room]:
-        return self.room.get_reserved_rooms()
+    def get_reserved_rooms(self, userID=None) -> List[Room]:
```
Comment (mclhtay) [inline HotelBooking/Controllers/room_controller.py]: `user_id`

Comment (mclhtay) [reply HotelBooking/Controllers/room_controller.py]: actually can you change all references of `user_id` to `customer_id` so it's consistent with the database attribute `customer_id` so it does not confuse others?

File: HotelBooking/Database/create_database.sql [hunk 16]
```diff
@@ -12,25 +13,50 @@ CREATE TABLE administrator
 CREATE TABLE room
 (
   room_id VARCHAR(20) NOT NULL PRIMARY KEY,
-  room_type TEXT CHECK(room_type IN ('SINGLE', 'DOUBLE', 'DELUXE', 'PRESIDENTIAL')) NOT NULL,
-  room_status TEXT CHECK(room_status IN('AVAILABLE', 'CHECKED-IN', 'RESERVED')) NOT NULL DEFAULT 'AVAILABLE',
-  customer_id INTEGER,
-  status_time TEXT,
-  FOREIGN KEY(customer_id) REFERENCES customer(customer_id)
+  room_type TEXT NOT NULL,
```
Comment (mclhtay) [inline HotelBooking/Database/create_database.sql]: why did you remove `CHECK(room_type IN ('SINGLE', 'DOUBLE', 'DELUXE', 'PRESIDENTIAL'))` ?

File: HotelBooking/Database/create_database.sql [hunk 17]
```diff
@@ -12,25 +13,50 @@ CREATE TABLE administrator
...
+  room_type TEXT  NOT NULL PRIMARY KEY,
+  room_price FLOAT NOT NULL
 );
 
+
 CREATE TABLE bill
 (
   bill_id INTEGER PRIMARY KEY,
-  bill_status TEXT CHECK(bill_status IN ('PAID', 'OUTSTANDING')) NOT NULL DEFAULT "OUTSTANDING",
+  bill_status TEXT CHECK(bill_status IN ('PAID', 'OUTSTANDING', 'CANCELED', 'refunded')) NOT NULL DEFAULT "OUTSTANDING",
```
Comment (mclhtay) [inline HotelBooking/Database/create_database.sql]: `REFUNDED` remain capital case for constants.

File: HotelBooking/Database/create_database.sql [hunk 18]
```diff
@@ -12,25 +13,50 @@ CREATE TABLE administrator
...
+  FOREIGN KEY(bill_id) REFERENCES bill(bill_id)
+);
+
+INSERT INTO roomType(room_type, room_price) VALUES("SINGLE", 100.00);
+INSERT INTO roomType(room_type, room_price) VALUES("DOUBLE", 150.00);
+INSERT INTO roomType(room_type, room_price) VALUES("DELUXE", 200.00);
+INSERT INTO roomType(room_type, room_price) VALUES("PRESIDENTIAL", 300.00);
+INSERT INTO customer(customer_id, customer_password) VALUES("bob", "aPassword");
+INSERT INTO customer(customer_id, customer_password) VALUES("cart", "aPassword2");
+INSERT INTO customer(customer_id, customer_password) VALUES("boob", "p3");
```
Comment (mclhtay) [inline HotelBooking/Database/create_database.sql]: Please use a different customer_id than `boob`, this is inappropriate.

File: HotelBooking/Models/bill.py [hunk 19]
```diff
@@ -24,6 +26,23 @@ def get_bill(self, bill_id: int) -> Bill:
         session.close()
         return bill
 
+    def get_available_bill(self, userID) -> List[Bill]:
```
Comment (mclhtay) [inline HotelBooking/Models/bill.py]: Plural. `customer_id`

File: HotelBooking/Models/bill.py [hunk 20]
```diff
@@ -24,6 +26,23 @@ def get_bill(self, bill_id: int) -> Bill:
...
+        engine = get_engine()
+        session = Session(engine)
+        statement = select(Bill).where(
+            Bill.bill_status == BILL_STATUS["OUTSTANDING"]).where(Bill.customer_id == userID)
+        bill = session.exec(statement).all()
+        session.close()
+        return bill
+
+    def get_all_bill(self, userID) -> List[Bill]:
```
Comment (mclhtay) [inline HotelBooking/Models/bill.py]: Plural & `customer_id`

File: HotelBooking/Models/bill.py [hunk 21]
```diff
@@ -45,3 +64,35 @@ def pay_bill(self, bill_id: int):
...
+    def refund_bill(self, bill_id: int):
+        bill = self.get_bill(bill_id)
+        engine = get_engine()
+        session = Session(engine)
+        bill.bill_status = BILL_STATUS["REFUNDED"]
+        session.add(bill)
+        session.commit()
+        session.close()
+
+    def modifyBill(self, bill_id: int, billAmount: float):
```
Comment (mclhtay) [inline HotelBooking/Models/bill.py]: `modify_bill`

Comment (mclhtay) [reply HotelBooking/Models/bill.py]: `bill_amount`

File: HotelBooking/Models/reservation.py [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+class Reservation(SQLModel, table=True):
+    reservation_id: int = Field(default=None, primary_key=True)
+    status: str
+    customer_id: Optional[str] = Field(foreign_key=Customer.customer_id)
+    room_id: Optional[str] = Field(foreign_key=Room.room_id)
+    bill_id: Optional[int] = Field(foreign_key=Bill.bill_id)
+    reservation_checkin_date: str
+    reservation_stay_date: int
+
+    def get_reservation_with_status(self, status, userID=None) -> List[Reservation]:
```
Comment (mclhtay) [inline HotelBooking/Models/reservation.py]: `get_reservations_with_status` & `customer_id`

File: HotelBooking/Models/reservation.py [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+        statement = select(Reservation).where(
+            Reservation.status == status)
+        if userID != None:
+            statement = statement.where(Reservation.customer_id == userID)
+
+        reservations = session.exec(statement).all()
+        session.close()
+        return reservations
+
+    def get_reservation_by_customer_id(self, userID) -> List[Reservation]:
```
Comment (mclhtay) [inline HotelBooking/Models/reservation.py]: ditto^

File: HotelBooking/Models/reservation.py [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+    def update_reservation_status(self, reservation_id, status):
+        reservation = self.get_reservation_by_id(reservation_id)
+        reservation.status = status
+        engine = get_engine()
+        session = Session(engine)
+        session.add(reservation)
+        session.commit()
+        session.close()
+
+    def update_reservation_start_date(self, reservation_id, newStartDate):
```
Comment (mclhtay) [inline HotelBooking/Models/reservation.py:68]: snake_case for parameter names, please.

File: HotelBooking/Models/reservation.py [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+        reservation = self.get_reservation_by_id(reservation_id)
+        reservation.reservation_checkin_date = newStartDate
+        engine = get_engine()
+        session = Session(engine)
+        session.add(reservation)
+        session.commit()
+        session.close()
+
+    def update_reservation_duration(self, reservation_id, newDuration):
```
Comment (mclhtay) [inline HotelBooking/Models/reservation.py:77]: ditto^

File: HotelBooking/Models/room.py [hunk 26]
```diff
@@ -23,14 +23,15 @@ class Room(SQLModel, table=True):
     room_id: Optional[str] = Field(primary_key=True)
     room_type: str
     room_status: str = Field(default="AVAILABLE")
-    status_time: Optional[str]
-    customer_id: Optional[int] = Field(foreign_key=Customer.customer_id)
+    customer_id: Optional[str] = Field(foreign_key=Customer.customer_id)
 
-    def get_room_with_status(self, room_status: ROOM_STATUS) -> List[Room]:
+    def get_room_with_status(self, room_status: ROOM_STATUS, userID=None) -> List[Room]:
```
Comment (mclhtay) [inline HotelBooking/Models/room.py]: plural & snake_case.

File: HotelBooking/Models/room.py [hunk 27]
```diff
@@ -61,8 +61,8 @@ def update_room_status(self, room_id: str, room_status: ROOM_STATUS, customer_id
     def get_checked_in_rooms(self) -> List[Room]:
         return self.get_room_with_status(ROOM_STATUS["CHECKED-IN"])
 
-    def get_reserved_rooms(self) -> List[Room]:
-        return self.get_room_with_status(ROOM_STATUS["RESERVED"])
+    def get_reserved_rooms(self, userID=None) -> List[Room]:
```
Comment (mclhtay) [inline HotelBooking/Models/room.py]: snake_case customer_id

File: HotelBooking/Models/roomType.py [hunk 28]
```diff
@@ -0,0 +1,17 @@
...
+from sqlmodel import Field, SQLModel, Session, select
+from HotelBooking.Models.utils import get_engine
+from typing import Optional
+
+
+class RoomType(SQLModel, table=True):
+    room_type: Optional[str] = Field(default=None, primary_key=True)
+    room_price: float
+
+    def get_Price(self, type: str) -> float:
```
Comment (mclhtay) [inline HotelBooking/Models/roomType.py]: snake_case ...

File: HotelBooking/Views/checked_in_rooms_view.py [hunk 29]
```diff
@@ -1,16 +1,21 @@
...
 
 PROMPT_KEY = {
     "OPERATIONS": "operations",
     "CHECK-IN": 'check-in',
     "ROOM-CHECK-IN": 'room-check-in',
     "CHECK-OUT": 'check-out',
-    "CUSTOMER": "customer"
+    "CUSTOMER": "customer",
+    "BACK": "back",
+    "LIST_RESERVATION": "listOutReservation"
```
Comment (mclhtay) [inline HotelBooking/Views/checked_in_rooms_view.py]: can you use consistent naming schemes? consider `LIST-RESERVATION`: `list-out-reservation`. Also, why is the key mapped to a value that is different?

File: HotelBooking/Views/checked_in_rooms_view.py [hunk 30]
```diff
@@ -67,12 +91,43 @@ def show(self):
         getattr(self, callable)()
 
     def check_in_customer(self):
-        check_in_choice = self.prompt_and_get_answer(
-            PROMPT_KEY['CHECK-IN'])
-        if check_in_choice:
-            self.reserved_check_in()
+        customer_id = self.prompt_and_get_answer(PROMPT_KEY['CUSTOMER'])
+        print(customer_id)
```
Comment (mclhtay) [inline HotelBooking/Views/checked_in_rooms_view.py]: Remove print. PyInquirer shows the last input.

File: HotelBooking/Views/checked_in_rooms_view.py [hunk 31]
```diff
@@ -94,58 +149,11 @@ def check_out_customer(self):
             if room_id != "Back":
                 self.room_controller.check_out_room(room_id)
                 print("\nSuccess!\n")
-            self.show()
-
-    def new_check_in(self):
```
Comment (mclhtay) [inline HotelBooking/Views/checked_in_rooms_view.py:99]: By removing this block, you're removing the ability to check in a walk-in customer who does not have a previous reservation. This is actually against the use-case diagram. Please add it back.

File: HotelBooking/Views/customer_view.py [hunk 32]
```diff
@@ -0,0 +1,74 @@
...
+        'message': "Customer operations",
+        'name': 'operations',
+        'choices': [
+        ]
+    }]
+}
+
+
+class CustomerView(View):
+    userID: str
```
Comment (mclhtay) [inline HotelBooking/Views/customer_view.py]: snake_case

File: HotelBooking/Views/modify_reservation_view.py [hunk 33]
```diff
@@ -0,0 +1,192 @@
...
+        self.reservation_controller = ReservationController()
+
+    def show(self):
+        operation = self.prompt_and_get_answer(PROMPT_KEY['OPERATIONS'])
+        callable = [operation_obj[1]
+                    for operation_obj in self.operation_options if operation_obj[0] == operation].pop()
+
+        getattr(self, callable)()
+
+    def changeStartDate(self):
```
Comment (mclhtay) [inline HotelBooking/Views/modify_reservation_view.py]: Please stay with consistent python name schemes.

File: HotelBooking/Views/pay_bill_view.py [hunk 34]
```diff
@@ -0,0 +1,140 @@
...
+        self.bill_controller = BillController()
+
+    def show(self):
+        operation = self.prompt_and_get_answer(PROMPT_KEY['OPERATIONS'])
+        callable = [operation_obj[1]
+                    for operation_obj in self.operation_options if operation_obj[0] == operation].pop()
+
+        getattr(self, callable)()
+
+    def payBill(self):
```
Comment (mclhtay) [inline HotelBooking/Views/pay_bill_view.py]: naming scheme.

File: HotelBooking/Views/reserved_history_view.py [hunk 35]
```diff
@@ -0,0 +1,127 @@
...
+        self.reservation_controller = ReservationController()
+
+    def show(self):
+        operation = self.prompt_and_get_answer(PROMPT_KEY['OPERATIONS'])
+        callable = [operation_obj[1]
+                    for operation_obj in self.operation_options if operation_obj[0] == operation].pop()
+
+        getattr(self, callable)()
+
+    def listStay(self):
```
Comment (mclhtay) [inline HotelBooking/Views/reserved_history_view.py]: naming scheme.

File: HotelBooking/Views/reserved_rooms_view.py [hunk 36]
```diff
@@ -1,14 +1,19 @@
...
+from HotelBooking.Models.roomType import RoomType
 PROMPT_KEY = {
     "OPERATIONS": 'operations',
     "RESERVE": 'reserve',
     "CANCEL": 'cancel',
-    "CUSTOMER": "customer"
+    "STARTDATE": "startDate",
+    "DURATION": "duration",
+    "BACK": "back",
+    "FINAL_CHECK": "finalCheck"
```
Comment (mclhtay) [inline HotelBooking/Views/reserved_rooms_view.py]: naming scheme.

File: HotelBooking/Views/reserved_rooms_view.py [hunk 37]
```diff
@@ -38,22 +43,58 @@
...
 
-    def __init__(self, history=[], caller=None) -> None:
+    def __init__(self, history=[], caller=None, userID=None, startDate=None, duration=None) -> None:
         super().__init__(history, caller)
         self.initiate_options()
         self.room_controller = RoomController()
+        self.userID = userID
+        self.startDate = startDate
+        self.duration = duration
+        self.roomType = RoomType()
```
Comment (mclhtay) [inline HotelBooking/Views/reserved_rooms_view.py]: A view should not be allowed to directly access a model.

**Review by mclhtay — CHANGES_REQUESTED** [review]
Unit tests for Models and Controllers?

Also next time, please do not request a PR of > 300 lines of code changes. #SmallPRs

File: HotelBooking/Views/checked_in_rooms_view.py (hunk 31 above)
Comment (vzhang23) [reply HotelBooking/Views/checked_in_rooms_view.py:99]: Please review the method "modify_reservation", I had delete the function to check-in customer who does not have a previous reservation, but I have added the funtion for the stuff to book/cancel a reservation for customer. If stuff want to check-in a customer who do not have reservation, he can simply book a reservation for that customer and than check him in

File: HotelBooking/Database/create_database.sql (hunk 16 above)
Comment (vzhang23) [reply HotelBooking/Database/create_database.sql]: because room-type do not have to be in 'SINGLE', 'DOUBLE', 'DELUXE', 'PRESIDENTIAL'

File: HotelBooking/Views/checked_in_rooms_view.py (hunk 31 above)
Comment (mclhtay) [reply HotelBooking/Views/checked_in_rooms_view.py:99]: Why not keep the check-in method exposed to the admin view and do the reserve & check-in in the exposed method behind the scene? In my opinion, it's better than asking the admin to perform 2 operations.

File: HotelBooking/Database/create_database.sql (hunk 16 above)
Comment (mclhtay) [reply HotelBooking/Database/create_database.sql]: then you should keep this check in the `roomType` table as it is being foreignly referrenced.

File: HotelBooking/Views/checked_in_rooms_view.py (hunk 31 above)
Comment (vzhang23) [reply HotelBooking/Views/checked_in_rooms_view.py:99]: If we are using the old way, it will not leave a record of the stay. (I need to track all the reservation history)

Comment (mclhtay) [reply HotelBooking/Views/checked_in_rooms_view.py:99]: I'm not saying we do the old way. I'm saying keep a method on the appropriate controller called `check_in_walk_in_customer` for `check_in_view`, and `check_in_walk_in_customer` will do:

1. make an immediate reservation 
2. modify this reservation to check-in status

Comment (vzhang23) [reply HotelBooking/Views/checked_in_rooms_view.py:99]: But It is also possable that hotel staff only want to make a reservation for a customer. I think it is better to leave it in two different functions

Comment (mclhtay) [reply HotelBooking/Views/checked_in_rooms_view.py:99]: I know that, but the hotel staff should be able to just "check-in" a walk-in customer directly, the controller will do the reservation modification behind the scenes. I am not saying to delete the reservation method, I'm saying to create the option of check-in flow for `check_in_walk_in_customer`.

Comment (mclhtay) [reply HotelBooking/Views/checked_in_rooms_view.py:99]: The old version, an admin could reserve a room for a customer AND check-in a customer without a reservation directly. In your current version, an admin can reserve a room for a customer, or check-in by first reserving and then modifying that reservation. I'm saying to create a new check-in for walk-in customers flow that asks the customer for the same parameters, but the method `check_in_walk_in_customer` would create a new reservation+modify reservation to check-in status and would save the admin a back and forth operation.

Comment (vzhang23) [reply HotelBooking/Views/checked_in_rooms_view.py:99]: If so, it is kind of redundant that a hotel stuff have the option to check-in customer, make reservation for customer, and there is also an option to check-in and make reservation.

Comment (mclhtay) [reply HotelBooking/Views/checked_in_rooms_view.py:99]: Your check-in customer is checking-in customers with a reservation. For an admin to check-in a walk-in customer, they have to navigate to reservation, then navigate back, and then checking that reservation in. 
Create an operation dedicated to serve walk-in customers and save that back and forth operation.

Comment (vzhang23) [reply HotelBooking/Views/checked_in_rooms_view.py:99]: sure, if you really want to...

Comment (mclhtay) [reply HotelBooking/Views/checked_in_rooms_view.py:99]: Actually, you don't have to add it if you don't want to. Sorry for being picky on this, I keep forgetting this is for school and not for work.

Comment (mclhtay) [reply HotelBooking/Views/checked_in_rooms_view.py:99]: We don't have real users of the app anyway, it is fine to be unclear in the UI I guess.

Comment (vzhang23) [reply HotelBooking/Views/checked_in_rooms_view.py:99]: just that I need to add a complete new view, which means that I also need to modify my diagram...

Comment (mclhtay) [reply HotelBooking/Views/checked_in_rooms_view.py:99]: NVM the request then (even though I don't see why you need to add a new view instead of just a new select option in check_in_rooms_view).

Comment (vzhang23) [reply HotelBooking/Views/checked_in_rooms_view.py:99]: Currently I am using the "reserved_rooms" view for both the customer and admin side, but since check-in is the functionality that only admin can have, I need to seprate the "reserved_rooms" for the customer and the admin.

Comment (mclhtay) [reply HotelBooking/Views/checked_in_rooms_view.py:99]: If it causes more breakage then forget it. Case closed.

**Comment by vzhang23** [comment]
all fixed

File: HotelBooking/Controllers/reservation_controller.py [hunk 38]
```diff
@@ -9,68 +10,66 @@
...
+        self.room.update_room_status(room_id, "RESERVED")
         r = self.room.get_room_by_id(room_id)
-        price = float(self.roomType.get_Price(r.room_type))
+        price = float(self.room_type.get_price(r.room_type))
         duration = int(duration)
         bill: Bill = self.bill_controller.create_bill(
             customer_id, price*duration)
-        self.reservation.create_reservation(
-            "OPEN", customer_id, room_id, bill.bill_id, startDate, duration)
+        print(status)
```
Comment (mclhtay) [inline HotelBooking/Controllers/reservation_controller.py]: Delete print

Comment (vzhang23) [reply HotelBooking/Controllers/reservation_controller.py]: deleted

File: HotelBooking/Views/reserved_rooms_admin_view.py [hunk 39]
```diff
@@ -0,0 +1,272 @@
...
+        'name': 'customer',
+    }],
+    "start_date": [{
+        'type': 'input',
+        'message': "Enter the start date",
+        'name': 'start_date',
+    }],
+    "duration": [{
+        'type': 'input',
+        'message': "Enter the amount of date you want to reserve",
```
Comment (mclhtay) [inline HotelBooking/Views/reserved_rooms_admin_view.py]: Change to "Enter the number of days you want to reserve"

**Review by mclhtay — CHANGES_REQUESTED** [review]
Can you `reset-db` to the default state? When I checked out, the db file contains manually modified values. We want to keep the db file in the repository as the default fresh database.

File: HotelBooking/Views/reserved_rooms_admin_view.py (hunk 39 above)
Comment (vzhang23) [reply HotelBooking/Views/reserved_rooms_admin_view.py]: fixed

**Comment by vzhang23** [comment]
> Can you `reset-db` to the default state? When I checked out, the db file contains manually modified values. We want to keep the db file in the repository as the default fresh database.

fixed

File: HotelBooking/Views/reserved_rooms_admin_view.py [hunk 40]
```diff
@@ -51,7 +51,7 @@
     }],
     "duration": [{
         'type': 'input',
-        'message': "Enter the amount of date you want to reserve",
```
Comment (mclhtay) [inline HotelBooking/Views/reserved_rooms_admin_view.py]: days not date

Comment (vzhang23) [reply HotelBooking/Views/reserved_rooms_admin_view.py]: fixed

**Review by mclhtay — APPROVED** [review]
Approval on the trust that you did manual UI testing.

Please remember to `squash and merge` and delete all your unused branches after the merge to maintain branch health.
